File: BaekJoon/sort.py
def insert_sort(A):
    for i in range(len(A)-1):
        for j in range(i+1,0,-1):
            print(i,j)
            if(A[j]<A[j-1]):
                A[j],A[j-1]=A[j-1],A[j]
            else:
                break;

def select_sort(A):
    for i in range(len(A)-1):
        min_index=i
        for j in range(i+1,len(A)):
            if(A[i]>A[j]):
                min_index=j
                A[min_index],A[i]=A[i],A[min_index]

def bubble_sort(A):
    for i in range(len(A)-1):
        for j in range(len(A)-i-1):
            if(A[j]>A[j+1]):
                A[j],A[j+1]=A[j+1],A[j]

def Quick_Sort(data):
    if(len(data)<2):
        return data
    else:
        pivot = data[0]
        left = [i for i in data[1:] if i<data[0]]
        right = [i for i in data[1:] if i>=data[0]]
        return Quick_Sort(left) + [pivot]+Quick_Sort(right)

def merge_sort(array):
    if(len(array)<=1):
         return array
    mid = len(array)//2
    left = merge_sort(array[:mid])
    right = merge_sort(array[mid:])
    i,j,k = 0 , 0 , 0

    while i<len(left) and j <len(right):
        if left[i] < right[j]:
            array[k] = left[i]
            i += 1
        else:
            array[k] = right[j]
            j += 1
        k += 1
    if i==len(left):
        while j< len(right):
            array[k] = right[j]
            j += 1
            k += 1
    elif j==len(right):
        while i< len(left):
            array[k] = left[i]
            i += 1
            k += 1
    return array

File: Algorithm/javascript/sort.js
const arr = [14, 38, 5, 44, 3];

function bubbleSort(arr) {
  for (let i = 0; i < arr.length; i++) {
    for (let j = 0; j < arr.length - i - 1; j++) {
      if (arr[j] > arr[j + 1]) {
        [arr[j], arr[j + 1]] = [arr[j + 1], arr[j]];
      }
    }
  }
  console.log(arr);
  return arr;
}

function selectSort(arr) {
  for (let i = 0; i < arr.length; i++) {
    let minIndex = i;
    for (let j = i + 1; j < arr.length; j++) {
      if (arr[minIndex] > arr[j]) {
        minIndex = j;
      }
    }
    [arr[minIndex], arr[i]] = [arr[i], arr[minIndex]];
  }
  console.log(arr);
  return arr;
}

function insertSort(arr) {
  for (let i = 1; i < arr.length; i++) {
    for (let j = i; j >= 0; j--) {
      if (arr[j - 1] > arr[j]) {
        [arr[j - 1], arr[j]] = [arr[j], arr[j - 1]];
      }
    }
  }
  console.log(arr);
  return arr;
}

function quickSort(arr) {
  if (arr.length <= 1) {
    return arr;
  }
  const left = [];
  const right = [];
  const mid = arr[0];
  for (let i = 1; i < arr.length; i++) {
    if (arr[i] <= mid) {
      left.push(arr[i]);
    } else {
      right.push(arr[i]);
    }
  }
  return quickSort(left).concat(mid, quickSort(right));
}

function mergeSort(arr) {
  if (arr.length <= 1) {
    return arr;
  }
  const mid = Math.floor(arr.length / 2);
  const left = arr.slice(0, mid);
  const right = arr.slice(mid);

  return merge(mergeSort(left), mergeSort(right));

  function merge(left, right) {
    const result = [];
    while (left.length && right.length) {
      if (left[0] < right[0]) {
        result.push(left.shift());
      } else {
        result.push(right.shift());
      }
    }
    return result.concat(left.slice(), right.slice());
  }
}

function heapSort(arr) {
  let len = arr.length;
  for (let i = Math.floor(len / 2) - 1; i >= 0; i--) {
    heapify(arr, i, len);
  }
  for (let i = len - 1; i > 0; i--) {
    [arr[0], arr[i]] = [arr[i], arr[0]];
    heapify(arr, 0, i);
  }
  return arr;
}

function heapify(arr, i, len) {
  let left = 2 * i + 1;
  let right = 2 * i + 2;
  let max = i;
  if (left < len && arr[left] > arr[max]) {
    max = left;
  }
  if (right < len && arr[right] > arr[max]) {
    max = right;
  }
  if (max !== i) {
    [arr[i], arr[max]] = [arr[max], arr[i]];
    heapify(arr, max, len);
  }
}

// bubbleSort(arr);
// selectSort(arr);
// insertSort(arr);
// console.log(quickSort(arr));
console.log(mergeSort(arr));
